Fixed PKCS5_PBKDF2_HMAC() to adhere to the documentation.

peterbudai opened the pull request:
The documentation of this function states that the password parameter
can be `NULL`. However, the implementation returns an error in this case
due to the inner workings of the `HMAC_Init_ex()` function.
With this change, `NULL` password will be treated as an empty string and
`PKCS5_PBKDF2_HMAC()` no longer fails on this input.

I have also added two new test cases that tests the handling of the
special values `NULL` and `-1` of the `password` and `passlen` parameters,
respectively.

kaduk commented on crypto/evp/p5_crpt2.c:
```diff
@@ -47,9 +48,10 @@ int PKCS5_PBKDF2_HMAC(const char *pass, int passlen,
+        pass = empty;
         passlen = 0;
-    else if (passlen == -1)
+    } else if (passlen == -1)
```
There might be some style tweaks needed, here.
OpenSSL style is usually to compare pointers against NULL, not to treat them as booleans.
I forget whether the rule about braces involves using braces for all branches of a conditional or none of them (as opposed to mixed, which is what this commit is doing).

richsalz commented:
Pointers should be explicitly compared against NULL, not via !.  If any arm of a conditional needs curly braces, they all get it.  Nested conditionals or almost anything other than a simple assignment would almost definitely push it to “curly brace the lot of them” territory.

peterbudai commented:
Thanks, I fixed the style problems you mentioned.

snhenson commented:
Could the tests be added to evptests.txt instead?

peterbudai commented:
There are two new test cases: one for zero length input and one for `NULL` input.
Unfortunately, EVP tests are not prepared for the latter, partly because of how `HMAC_Init_ex()` API is contructed, which also caused the incorrect behavior being fixed in this PR. See comment at [line 132 of `evp_test.c`](https://github.com/openssl/openssl/blob/master/test/evp_test.c#L132).

So while it is possible to add the first test case with ease, adding the other requires more changes in `evp_tests.c` as well to make it recognize NULL as valid input. Of course I'd do it if you think it was useful.

peterbudai commented:
Disregard my previous comment :) I found an obvious way to specifiy `NULL` password in `evptests.txt`, so I have added the two test cases to `evptest.txt` as well as you requested.

levitte commented:
Now that you have the test cases in `test/evptest.txt`, is there a reason why they are also still in `test/p5_crpt2_test.c`?

snhenson commented:
I think it would make sense to support the string NULL as valid for buffer values as well as not specifying it at all. That would be a simple change to test_bin.

peterbudai commented:
I removed the new test cases from `p5_crpt2_test.c` and updated `evp_test.c` to also recognize `NULL` as value. However, I`m still struggling why those tests won't succeed on AppVeyor. I've compiled and run the test on a Windows machine as well, with the exact same compiler and config parameters and it ran all fine on my computer.

peterbudai commented:
I was finally able to reproduce the failing tests on my Windows machine. It is related to the `autocrlf` setting of Git. If `evptests.txt` are checked out with CRLF line endings on Windows (on my computer), the tests are fine. On the other side if the file is checked out with LF line endings, the public/private key parsing seems to be messing up the line counting, and the X25519 test vectors at the end of the file will fail. Though the failing tests are unrelated to the new PBKDF2 test vectors I added (they run successfully), somehow the issue only arises when those test cases are present.

Now that I am able to debug on my own computer, I`m on finding the cause and fixing it.

peterbudai commented:
It turned out that the test failures where caused by the faulty `ftell()` implementation within the Microsoft C Runtime when files with UNIX-style line endings are opened in text mode (`"r"` mode string). Here is the related [Microsoft Connect issue](https://connect.microsoft.com/VisualStudio/feedback/details/3107461), basically `ftell` (and thus `BIO_tell`) gives bogus file positions in this case that are several bytes away from the exact location, and because of this `BIO_seek` will jump to an incorrect location after reading a public/private key in `process_test()`.

When `evptest.txt` reached a specific file size (that happened to coincide with me adding 6 more PBKDF2 test cases) the file position differences caused by the faulty `ftell()` resulted in one of the X25519 private keys not being parsed and thus the related tests failing.

The workaround was to open the test file in binary mode in `evp_test.c`. Now the tests succeed on every platform.

richsalz commented:
@peterbudai can you sign our cla?  and rebase?  thanks.

peterbudai commented:
Sure, I sent signed ICLA today for buc@peterbudai.eu via email and rebased / cleaned up the pull request.

levitte commented:
Hi @peterbudai...  sorry for the slow response.  I can't find any ICLA from you among my emails or in our CLA db.  Did you send it and where?

peterbudai commented:
Hi @levitte, yes I did send it to legal at opensslfoundation dot org, and I resent it a few seconds ago, just to be sure.

peterbudai commented:
All green, finally :) If you need something else from me to have this merged, please let me now.

richsalz commented:
fa013b6 in master and 8d049ed in 1.1.0  Thanks for your patience and, more importantly, the patch! :)
